libvast/src/arrow_table_slice.cpp

```diff
@@ -1141,4 +1146,171 @@ template class arrow_table_slice<fbs::table_slice::arrow::v0>;
 template class arrow_table_slice<fbs::table_slice::arrow::v1>;
 template class arrow_table_slice<fbs::table_slice::arrow::experimental>;
 
+// -- utility functions --------------------------------------------------------
+
+namespace {
+
+std::shared_ptr<arrow::Array>
+convert_subnet_array(const arrow::FixedSizeBinaryArray& arr) {
+  auto* pool = arrow::default_memory_pool();
+  auto subnet_builder = subnet_type::make_arrow_builder(pool);
+  auto& address_builder = subnet_builder->address_builder();
+  auto& length_builder = subnet_builder->length_builder();
+  if (!address_builder.Resize(arr.length()).ok()
+      || !length_builder.Resize(arr.length()).ok()
+      || !subnet_builder->Resize(arr.length()).ok())
+    die("unable to resize builders to target capacity");
+  for (int row = 0; row < arr.length(); ++row) {
+    if (arr.IsNull(row)) {
+      if (!subnet_builder->AppendNull().ok())
+        die("append null value to subnet builder failed");
+    } else {
+      auto span = std::span<const uint8_t, 17>{arr.GetValue(row), 17};
+      if (!(subnet_builder->Append().ok()
+            && address_builder.Append(arr.GetValue(row)).ok()
+            && length_builder.Append(span[16]).ok()))
+        die("rebuilding of subnet struct array failed");
+    }
+  }
+  if (auto result = subnet_builder->Finish(); result.ok())
+    return *result;
+  die("failed to finish Arrow subnet column builder");
+}
+
+// TODO: this is copied from
+// experimental_table_slice_builder/enum_column_builder, can this live somewhere
+// shared?
+std::shared_ptr<arrow::Array> make_dict_array(const enumeration_type& et) {
+  arrow::StringBuilder dict_builder{};
+  for (const auto& f : et.fields())
+    if (!dict_builder.Append(std::string{f.name}).ok())
+      die("failed to build Arrow enum field array");
+  if (auto array = dict_builder.Finish(); array.ok()) {
+    return *array;
+  }
+  die("failed to finish Arrow enum field array");
+}
+
+/// Converts an array representing a single column from a previous arrow
+/// format to `arrow::experimental`.
+std::shared_ptr<arrow::Array>
+convert_column(const std::shared_ptr<arrow::Array>& arr, const type& t) {
+  auto f = detail::overload{
+    [&](const auto& t) -> std::shared_ptr<arrow::Array> {
+      VAST_ASSERT(arr->type()->Equals(t.to_arrow_type()));
+      return arr;
+    },
+    [&](const record_type& rt) -> std::shared_ptr<arrow::Array> {
+      // this case handles VAST type `list<record>`
+      const auto& sa = static_pointer_cast<arrow::StructArray>(arr);
+      arrow::ArrayVector children{};
+      std::vector<std::string> field_names{};
+      children.reserve(rt.num_fields());
+      field_names.reserve(rt.num_fields());
+      int index = 0;
+      for (const auto& f : rt.fields()) {
+        const auto& arr = sa->field(index++);
```

Comment:
I think `detail::zip` makes this easier to read.

Reply:
I don't think that works because one is a `vector`, the other a generator:

```
vast/libvast/src/arrow_table_slice.cpp:1211:72: error: no matching function for call to ‘zip(vast::detail::generator<vast::record_type::field_view>, const ArrayVector&)’
```

Do you have a suggestion on how to fix that?

Reply:
That's probably just overly constrained, because I don't understand why `zip` wouldn't work with input iterators. Maybe just leave it as-is then, not worth spending time on now imo.